### PR TITLE
[datadogexporter] Use resource attributes for metadata and generated metrics

### PR DESCRIPTION
**Pull request by mx-psi**
**Description:** 

- Set hostname to running metrics from resource attributes if available
- Use resource metadata to complete host metadata
- Improve host metadata on cloud environments (AWS and GCP)
- Add tests for new metadata behavior

**Link to tracking Issue:** Relates to #1379
**Testing:** Added unit tests. Tested in e2e environment.

- Set up the resource detection processor with the environment detector
- Use it to add a `datadog.host.name` to something like `my-custom-hostname`
- Test that it outputs metrics and metadata with `my-custom-hostname`

Testing in AWS and GCP should bring a subset of the information that the Datadog Agent provides on its latest version (7.25)

**Documentation:** Documented new flag

### CODE REVIEW

File: exporter/datadogexporter/metadata/ec2/ec2.go [hunk 1]
```diff
@@ -23,11 +24,15 @@ import (
 	"go.uber.org/zap"
 )
 
-var defaultPrefixes = [3]string{"ip-", "domu", "ec2amaz-"}
+var (
+	defaultPrefixes = [3]string{"ip-", "domu", "ec2amaz-"}
+	ec2TagPrefix    = "ec2.tag."
```
Comment (mx-psi) [inline exporter/datadogexporter/metadata/ec2/ec2.go:29]: This needs to match the prefix defined in #1899, see [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/processor/resourcedetectionprocessor/internal/aws/ec2/ec2.go#L35)

File: exporter/datadogexporter/metadata/gcp/gcp.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	if hostName, ok := attrs.Get(conventions.AttributeHostName); ok {
+		return hostName.StringVal(), true
+	}
+
+	return "", false
+}
+
+// HostInfoFromAttributes gets GCP host info from attributes following
+// OpenTelemetry semantic conventions
+func HostInfoFromAttributes(attrs pdata.AttributeMap) (hostInfo *HostInfo) {
```
Comment (mx-psi) [inline exporter/datadogexporter/metadata/gcp/gcp.go:40]: This function should follow the same logic as this code from the Datadog Agent https://github.com/DataDog/datadog-agent/blob/491c309e374ed5c7f1b385b347d5f9b5ac72c2b6/pkg/util/gce/gce_tags.go#L71-L101

File: exporter/datadogexporter/metadata/metadata.go [hunk 3]
```diff
@@ -58,6 +61,14 @@ type HostMetadata struct {
 type HostTags struct {
 	// OTel are host tags set in the configuration
 	OTel []string `json:"otel,omitempty"`
+
+	// System are host tags from EC2.
+	// For compatibility with the Datadog Agent we use the
+	// system host tags to send EC2 host tags
```
Comment (mx-psi) [inline exporter/datadogexporter/metadata/metadata.go:67]: This can be checked here https://github.com/DataDog/datadog-agent/blob/491c309e374ed5c7f1b385b347d5f9b5ac72c2b6/pkg/metadata/host/host_tags.go#L97-L139

I am a bit worried that this will cause trouble if an Agent and a Collector are running at the same time

File: exporter/datadogexporter/factory.go [hunk 4]
```diff
@@ -99,23 +100,24 @@ func createMetricsExporter(
 		return nil, err
 	}
 
+	ctx, cancel := context.WithCancel(ctx)
 	var pushMetricsFn exporterhelper.PushMetrics
 
 	if cfg.OnlyMetadata {
-		pushMetricsFn = func(context.Context, pdata.Metrics) (int, error) {
-			// if only sending metadata ignore all metrics
+		pushMetricsFn = func(_ context.Context, md pdata.Metrics) (int, error) {
```
Comment (KSerrania) [inline exporter/datadogexporter/factory.go:107]: Codecov is complaining, probably because this function (and the one for traces) never gets called in tests. We could add a test that makes sure this is called, with a mock server that receives the metadata payload.

File: exporter/datadogexporter/factory.go [hunk 5]
```diff
@@ -99,23 +100,24 @@ func createMetricsExporter(
...
+			// only sending metadata use only metrics
+			once := cfg.OnceMetadata()
+			once.Do(func() {
+				attrs := pdata.NewAttributeMap()
+				if md.ResourceMetrics().Len() > 0 {
```
Comment (KSerrania) [inline exporter/datadogexporter/factory.go:112]: There may be one weird edge case here: what if, during the invocation of `pushMetricsFn`, there may be sometimes (but not always) no `ResourceMetrics` in `md`?. This could lead to some weird behavior where upon restarting the collector, metadata may change depending on whether  the first`md` object passed to the exporter contains reosurce metrics or not, which may be confusing.

I don't know if there's a good solution to this though. You could wait until you find a `md` with resource metrics, but that means metadata isn't sent as fast as possible, which would add its own set of problems.

File: exporter/datadogexporter/metadata/gcp/gcp.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+// HostnameFromAttributes gets a valid hostname from labels
+// if available
+func HostnameFromAttributes(attrs pdata.AttributeMap) (string, bool) {
+	if hostName, ok := attrs.Get(conventions.AttributeHostName); ok {
+		return hostName.StringVal(), true
+	}
+
+	return "", false
```
Comment (KSerrania) [inline exporter/datadogexporter/metadata/gcp/gcp.go:35]: This Codecov warning can be eliminated by adding a test where we call `HostnameFromAttributes` with an empty attribute map, and check that the return is `"", false`.

File: exporter/datadogexporter/metadata/metadata.go [hunk 7]
```diff
@@ -81,25 +92,58 @@ type Meta struct {
...
+		hm.Meta.Hostname = hostname
+	}
+
+	// AWS EC2 resource metadata
+	cloudProvider, ok := attrs.Get(conventions.AttributeCloudProvider)
+	if ok && cloudProvider.StringVal() == conventions.AttributeCloudProviderAWS {
+		ec2HostInfo := ec2.HostInfoFromAttributes(attrs)
+		hm.Meta.InstanceID = ec2HostInfo.InstanceID
+		hm.Meta.EC2Hostname = ec2HostInfo.EC2Hostname
+		hm.Tags.System = ec2HostInfo.EC2Tags
```
Comment (KSerrania) [inline exporter/datadogexporter/metadata/metadata.go:111]: Just checking, does that mean these tags will be shown as coming from the Datadog Agent, in the infrastructure list (since there's no difference between host tags and EC2 tags)?

File: exporter/datadogexporter/metrics_translator.go [hunk 8]
```diff
@@ -261,12 +263,21 @@ func mapMetrics(cfg config.MetricsConfig, prevPts *ttlmap.TTLMap, md pdata.Metri
...
+		}
+		series = append(series, runningMetric...)
+
 		ilms := rm.InstrumentationLibraryMetrics()
 		for j := 0; j < ilms.Len(); j++ {
 			ilm := ilms.At(j)
-			metrics := ilm.Metrics()
-			for k := 0; k < metrics.Len(); k++ {
-				md := metrics.At(k)
+			metricsArray := ilm.Metrics()
```
Comment (KSerrania) [inline exporter/datadogexporter/metrics_translator.go:278]: Note: these codecov warnings are due to the lack of unit tests for `mapMetrics`, they'll be fixed in another PR

**Review by KSerrania — COMMENTED** [review]
The code change looks good to me overall, I left some questions & notes about tests

File: exporter/datadogexporter/factory.go (hunk 5 above)
Comment (mx-psi) [reply exporter/datadogexporter/factory.go:112]: I think this can't happen if the pipeline configuration doesn't change. The processors are initialized before the receivers and they gather metadata at startup, if they fail to do so the Collector process stops. So, (if my understanding of the code is correct) it can either be that the Collector fails to start or that the metadata is gathered correctly and applied to every payload.

File: exporter/datadogexporter/factory.go (hunk 4 above)
Comment (mx-psi) [reply exporter/datadogexporter/factory.go:107]: Done!

File: exporter/datadogexporter/metadata/metadata.go (hunk 7 above)
Comment (mx-psi) [reply exporter/datadogexporter/metadata/metadata.go]: I changed it to add them to the `OTel` field since I think it makes more sense

**Review by KSerrania — APPROVED** [review]
Approving on behalf of Datadog.

**Comment by mx-psi** [comment]
Marking as ready for review; it was approved by someone from Datadog. Note that CodeCov warnings due to lack of tests for `mapMetrics` will be covered in a future PR.

**Review by bogdandrutu — APPROVED** [review]
